python/paddle/fluid/contrib/slim/quantization/dyquant/dygraph_quantization.py

```diff
@@ -0,0 +1,126 @@
+        super(DygraphQuantAware, self).__init__()
+        self._weight_bits = weight_bits
+        self._activation_bits = activation_bits
+        self._moving_rate = moving_rate
+
+        quant_type = {'abs_max', 'moving_average_abs_max'}
+        if activation_quantize_type not in quant_type:
+            raise ValueError(
+                "Unknown activation_quantize_type : '%s'. It can only be "
+                "'abs_max' or 'moving_average_abs_max' now." %
+                (str(activation_quantize_type)))
+        if weight_quantize_type not in quant_type:
+            raise ValueError(
+                "Unknown weight_quantize_type: '%s'. It can only be "
+                "'abs_max' or 'moving_average_abs_max' now." %
+                (str(weight_quantize_type)))
+        self._activation_quantize_type = activation_quantize_type
+        self._weight_quantize_type = weight_quantize_type
+
+        self._quant_layers_map = {'Conv2D': Conv2D, 'Linear': Linear}
+        self._translator = dygraph.ProgramTranslator()
+        self._quantizable_layer_type = tuple(
+            self._quant_layers_map[layer]
+            if layer in self._quant_layers_map else layer
+            for layer in quantizable_layer_type)
+        for layer in self._quantizable_layer_type:
+            assert not isinstance(
+                layer, str), "{} is unspported to be quantized.".format(layer)
+
+    def prepare(self):
+        self._translator.enable_declarative = False
+
+    def quantize(self, model):
+        for name, layer in model.named_sublayers():
+            if not isinstance(layer, self._quantizable_layer_type):
+                continue
+
+            scopes = name.split('.')
+            target = scopes[-1]
+            obj = model
+            parent = model
+            for i in range(len(scopes) - 1):
+                obj = getattr(parent, scopes[i])
+                parent = obj
+
+            quant_layer = self._get_quantized_counterpart(layer)
+            setattr(obj, target, quant_layer)
+
+    def save_infer_quant_model(self,
+                               dirname,
+                               model,
+                               input_shape,
+                               input_dtype='float32',
+                               feed=None,
+                               fetch=None,
+                               append_batch_size=True):
+        with dygraph.guard():
+            self._translator.enable_declarative = True
+            model.eval()
+            raw_data = np.random.random(input_shape)
+            input_data = raw_data[np.newaxis, :].astype(
+                input_dtype) if append_batch_size else raw_data.astype(
+                    input_dtype)
+            input_var = dygraph.to_variable(input_data)
+            out = model(input_var)
+        self._translator.save_inference_model(dirname, feed, fetch)
+
+    def _get_quantized_counterpart(self, layer):
```

Comment:
In the function, can we get the `counterpart` of `layer`, i.e. corresponding quant_layer  of a given `layer` directly? I guess we do not need to construct `quant_layers` every time and get the index of `layer`?

Reply:
This is just the same as `switch` clause. Otherwise, there will be too many `if ... elif...` clause.